core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -325,12 +325,18 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
 
   override def getListing(): Iterator[ApplicationInfo] = {
     // Return the listing in end time descending order.
-    listing.view(classOf[ApplicationInfoWrapper])
+    // SPARK-38896: tryWithResource cannot be used here
+    // because this method needs to return an `Iterator`.
+    val closeableIter = listing.view(classOf[ApplicationInfoWrapper])
       .index("endTime")
       .reverse()
-      .iterator()
-      .asScala
-      .map(_.toApplicationInfo())
+      .closeableIterator()
+    val dataIter = closeableIter.asScala.map(_.toApplicationInfo())
+    new Iterator[ApplicationInfo] with Closeable {
+      override def hasNext: Boolean = dataIter.hasNext
+      override def next(): ApplicationInfo = dataIter.next()
+      override def close(): Unit = closeableIter.close()
+    }
```

Comment:
Since not all usages of `getListing()` are draining the iterator, add a `finalize` as well ?

Reply:
`finalize` or `finally`?

Reply:
If `finalize`, there is no need to change here. But I don't think it should be used `finalize()` anymore. It's not a good practice, and it has been identified as forRemoval=true after [JEP 421](https://openjdk.java.net/jeps/421)

If use `finally` here to close `LevelDB/RocksDBIterator`, the caller will not be able to consume data.

Reply:
finalize() isn't wrong per se; relying on it can be. Given the context here, I think you would want to include it.

Reply:
Thank you for your help  @srowen

Reply:
Or we can return
```
KVUtils.mapToSeq(listing.view(classOf[ApplicationInfoWrapper])
      .index("endTime")
      .reverse())(_.toApplicationInfo()).iterator
```
This seems simpler, and the caller doesn't need to care about closing `LevelDB/RocksDBIterator`, but it may use more memory